samples/99_zil_interpreter/app/main.rb

```diff
@@ -7,39 +7,106 @@
 
 def tick(args)
   setup(args) if args.tick_count.zero?
-  # TODO:
-  # - Render history
-  # - Collect input
-  # - Call send_input with input when pressing Enter
+  # Render history
+  render_state(args)
+  
+  # Collect input
+  handle_input(args)
+  
+  # Call send_input when pressing Enter
+  if args.inputs.keyboard.key_down.enter && args.state.input != ''
+    send_input(args, args.state.input)
+    args.state.input = ''
+  end
+  
   $gtk.request_quit unless $interpreter.alive?
 end
 
 def setup(args)
+  args.state.text_color = [240, 255, 255]
+  args.state.bg_color = [0, 0, 0]
+  args.state.input = ''
+  args.state.text_history = []
+  args.state.current_line = 0
   context = build_zil_context(args)
   args.state.zil_context = context
   $interpreter = Fiber.new {
     context.globals[:GO].call [], context
   }
   $interpreter.resume # Initial processing until first Fiber.yield
-  process_outputs args, context.outputs # Process welcome message if existing
+  process_outputs args, ' ', context.outputs # Process welcome message if existing
+  context.outputs.clear
 
   # TODO:
   # Add other setup if necessary
 end
 
-# Call this method with the input after pressing enter
+def handle_input(args)
+  if args.inputs.keyboard.key_down.backspace
+    args.state.input.chop!
+  else
+    if args.inputs.text[0]
+      args.state.input << args.inputs.text[0]
+    end
+  end
+
+  # Scrolling
+  if args.inputs.mouse.wheel
+    args.state.current_line += args.inputs.mouse.wheel.y
+
+    if args.inputs.mouse.wheel.y > 0
+      args.state.current_line = [args.state.current_line, (args.state.text_history.length - 33)].min
+    end
+    args.state.current_line = [args.state.current_line, 0].max
+
+  # Jump to the present
+  elsif args.inputs.keyboard.key_down.escape
+    args.state.current_line = 0
+  end
+end
+
+# Called with the input after pressing enter
 def send_input(args, input)
   $interpreter.resume input
 
   context = args.state.zil_context
-  process_outputs(args, context.outputs)
+  process_outputs(args, input, context.outputs)
   context.outputs.clear
 end
 
-def process_outputs(args, outputs)
-  # TODO:
-  # Replace with actual processing, e.g. adding it to the history of output lines etc
-  outputs.each do |line|
-    puts line
+def process_outputs(args, input, outputs)
+  unless args.tick_count.zero?
+    args.state.text_history << " "
+    args.state.text_history << "> #{input}"
+  end
+  args.state.text_history += outputs
+end
+
+def render_state(args)
+  args.outputs.background_color = args.state.bg_color
+  
+  # Player input
+  input_line = "> #{args.state.input}"
+  input_line << "_" if (args.tick_count / 32) % 2 < 1 # Blinky underscore like in all the old computers!
```

Comment:
```suggestion
  input_line << "_" if (args.tick_count / 32).mod_zero? 2 # Blinky underscore like in all the old computers!
```

DragonRuby offers `mod_zero?` which makes writing logic like this shorter :)

Reply:
Ah, it's a bit of number trickery... I need to use `< 1` here because I am neglecting to convert the number to an integer. `% 2 == 0` or `.mod_zero? 2` both are true for only one frame out of 32, instead of 16 of 32. However, `.round.mod_zero? 2` does work as intended

Reply:
(Alternatively, since >>/<< only returns integers, I could do `(args.tick_count >> 5).mod_zero? 2` but that seems _very_ opaque)